readthedocs/search/parsers.py

```diff
@@ -133,28 +149,102 @@ def _parse_sections(self, title, body):
             )
             if content:
                 yield {
-                    'id': '',
-                    'title': title,
-                    'content': content,
+                    "id": "",
+                    "title": document_title,
+                    "content": content,
                 }
         except Exception as e:
-            log.info('Unable to index section', section=str(e))
+            log.info("Unable to index section", section=str(e))
 
         # Index content from h1 to h6 headers.
-        for head_level in range(1, 7):
-            tags = body.css(f'h{head_level}')
-            for tag in tags:
+        for section in [body.css(f"h{h}") for h in range(1, 7)]:
+            for tag in section:
                 try:
-                    title, id = self._parse_section_title(tag)
+                    title, _id = self._parse_section_title(tag)
                     next_tag = self._get_header_container(tag).next
                     content, _ = self._parse_section_content(next_tag, depth=2)
                     yield {
-                        'id': id,
-                        'title': title,
-                        'content': content,
+                        "id": _id,
+                        "title": title,
+                        "content": content,
                     }
-                except Exception as e:
-                    log.info('Unable to index section.', section=str(e))
+                except Exception:
+                    log.info("Unable to index section.", exc_info=True)
+
+    def _parse_dls(self, body):
+
+        # All terms in <dl>s are treated as sections.
+        # We traverse by <dl> - traversing by <dt> has shown in experiments to render a
+        # different traversal order, which could make the tests more unstable.
+        dls = body.css("dl")
+
+        for dl in dls:
+
+            # Hack: Identify a :host() without using :host() selector
+            # We would perhaps like to have written '> dt' but the syntax isn't allowed.
+            # We also cannot use :host.
+            # So in order to ensure that we are only selecting the <dt> elements that are direct
+            # descendents of the traversed <dl>, we give it a unique ID.
+            random_uuid = uuid4().hex
+            # Here, it's okay to manipulate the ID of the <dl> because we're not going to use it.
+            # Note: It didn't work to use custom node attributes nor :has().
+            dl.attrs["id"] = random_uuid
+            # Select all dts with id defined
+            dts = dl.css(f'dl#{random_uuid} > dt[id]:not([id=""])')
```

Comment:
The `>` selector is basically doing all immediate children, so you can just use `dl.iter()` to iterate over all immediate children, and manually doing the filter of dts with ID. Not as fancy as the CSS selector, but I think it's more explicit than assigning the random ID.

```python
dts = (node for node in dl.iter() if node.tag == 'dt' and node.id)
```

Reply:
That works + and yes definitely much less code, the css selector dogma was silly :+1: